core/match_pattern.py

```diff
@@ -62,44 +62,99 @@ def compute_child_table(suf_tab, lcp_tab):
     return child_table
 
 
-def get_child_intervals(i, j, child_tab: list[Child]) -> list[Tuple[int, int]]:
-    intervals = []
+def is_interval_eligible(a: str, i: int, lcp: int, s: str, suf_tab: list[int]) -> bool:
+    idx_to_hit = suf_tab[i]
+
+    return s[idx_to_hit + lcp] == a
+
+
+def get_lcp(i: int, j: int, child_tab: list[Child], lcp_tab: list[int]):
+    assert i > -1 and j <= len(child_tab)
+
+    # root interval
+    if i == 0 and j == len(child_tab):
+        return 0
+    elif len(child_tab) > j > child_tab[j].up > i:
+        return lcp_tab[child_tab[j].up]
+    else:
+        return lcp_tab[child_tab[i].down]
+
+
+def get_interval(i: int, j: int, a: str, s: str, child_tab: list[Child],
+                 lcp_tab: list[int], suf_tab: list[int]) -> Tuple[int, int]:
     i_1: int = -1
     i_2: int = -1
 
+    assert i > -1 and j <= len(child_tab)
+
+    lcp = get_lcp(i, j, child_tab, lcp_tab)
+
     if i == 0 and j == len(child_tab):
         i_1 = child_tab[i].next_l_index
-    elif (j + 1 < len(child_tab)) and i < child_tab[j + 1].up <= j:
-        i_1 = child_tab[j + 1].up
+    elif len(child_tab) > j > child_tab[j].up > i:
+        i_1 = child_tab[j].up
     else:
         i_1 = child_tab[i].down
 
     # has no children
     if i_1 == -1:
-        return intervals
+        return -1, -1
 
-    intervals.append((i, i_1 - 1))
+    # runs in infinite loop for some test cases, so still returning (lb, rb - 1) instead of (lb, rb)
+    if is_interval_eligible(a, i, lcp, s, suf_tab):
+        return i, i_1 - 1
 
     while child_tab[i_1].next_l_index != -1:
         i_2 = child_tab[i_1].next_l_index
 
-        intervals.append((i_1, i_2 - 1))
+        if is_interval_eligible(a, i_1, lcp, s, suf_tab):
+            return i_1, i_2 - 1
 
         i_1 = i_2
 
-    intervals.append((i_1, j))
+    if is_interval_eligible(a, i_1, lcp, s, suf_tab):
+        return i_1, j
 
-    return intervals
+    # represents undefined
+    return -1, -1
 
 
-def get_lcp(i, j, child_table: list[Child], lcp_tab):
-    if (i < child_table[j + 1].up) and (child_table[j + 1].up <= j):
-        return lcp_tab[child_table[j + 1].up]
-    else:
-        return lcp_tab[child_table[i].down]
+def match_pattern(s: str, p: str, child_tab: list[Child], lcp_tab: list[int], suf_tab: list[int]) -> bool:
+    # considering empty pattern to be a valid string; returns false if the string s is non-empty
+    if len(p) == 0:
+        # len(s) == 1 is empty because we consider sentinel char $ at the end
+        return len(s) == 1
+
+    c: int = 0
+    pattern_found: bool = True
+    (n, m) = (len(s), len(p))
+    (i, j) = get_interval(0, n, p[c], s, child_tab, lcp_tab, suf_tab)
+
+    while i != -1 and j != -1 and c < m and pattern_found is True:
```

Comment:
The `i != -1 and j != -1` is ugly, and ugly code almost always indicates that someone didn't think hard enough about what they are trying to achieve.

I'm guessing it's from the paper and a way of indicating a failed search, but a failed search is already handled in two separated ways, what the remaining interval is and `pattern_found`. Right now, you really have three ways to detect if you should continue, and that is (at least) two too many.

If `not pattern_found` you want to bail out, because the pattern isn't there.
If `i == j`, then you have an empty interval to look in, so you should also bail.
If `i == -1` or `j == -1`, it probably means that some lookup function returns an error that it shouldn't return. There is no way that you shouldn't get child intervals when you ask for them, and you should never ask for child intervals on a leaf (and those are the intervals where `(i, j) == (i, i+1)`.

This procedure, as far as I can tell, is a recursive one. You have an interval, `(i, j)`. If you hit an empty interval, you don't have a match. If you hit a singleton you need to check it, and otherwise you need to match against the first string in an interval. You are interested in how much you already match, and if that reaches `len(p)` you are done. If you hit an empty interval along the way, or if you cannot match down a string, you should return that you don't have a match.

Imagine that we have a function for computing the length of a prefix. Something like this:

```python
def prefix(x: str, off_x: int,
           y: str, off_y: int,
           max: int | None = None) -> int:
    """
    Return how much of x[off_x:off_x+max] matches y[off_y:off_y+max].

    >>> prefix("foo", 0, "foobar", 0)
    3
    >>> prefix("bar", 0, "foobar", 3)
    3
    >>> prefix("baz", 0, "foobar", 3)
    2
    >>> prefix("barbaz", 0, "foobarbaz", 3, 2)
    2
    """
    max = max or min(len(x) - off_x, len(y) - off_y)
    for prefix, i in enumerate(range(max)):
        if x[off_x + i] != y[off_y + i]:
            return prefix
    return max
```

Then the search function could look something like this:

```python
# We will get either an interval or None
SearchResult = tuple[int, int] | None


def search(i: int, j: int, matched: int,
           p: str, x: str,
           sa: list[int]) -> SearchResult:
    """
    Search the interal (i, j) for p.

    We already know we match `matched`.
    There will be other arguments for additional data structures
    but I didn't bother guessing at their names.
    """
    while matched < len(p):
        if i == j:  # Empty interval means no result
            return None

        if i + 1 == j:  # singleton interval
            # If we don't have a full match now we never will
            matched += prefix(p, matched, x, sa[i] + matched)
            return i, j if len(p) == matched else None

        # (i, j) is an lcp-interval
        lcp = get_lcp(i, j)  # however you get the shared prefix of [i,j]
        to_match = min(len(p) - matched, lcp)
        match = prefix(p, matched, x, sa[i] + matched, to_match)
        if match < to_match:
            return None  # We cannot match

        matched += match
        if matched < len(p):
            # recurse with sub-interval that starts with letter p[matched]
            i, j = get_sub_interval(i, j, p, matched)

    return i, j
```

It is still far from pretty (but I don't have time to figure out a pretty version), but there is one and only one test condition in the loop, and I explicitly terminate when I can see that I cannot get any further.

This version is not much simpler, that would require more work, but it isn't complicated to see when it terminates at least.




Oh, and `pattern_found is True` is now how you write `pattern_found`. Booleans work as booleans, you don't need to check `if True is True` or `if False is True`. Just use the boolean value, it is what it is there for.